### PR TITLE
fix(common/core/web): Fixes engine-level context tests, adds notany cases

### LINKED ISSUE #3841: chore(web): Add unit test to validate notany() fix in #3817
This test seemed to be backwards. It could be dealt with as a separate PR because it's a separate bug, but I'm reasonably comfortable with rolling it in with this fix.

```
              } else if(r.n && (result || context[i] === NUL_CONTEXT)) {
```

_Originally posted by @mcdurdin in https://github.com/keymanapp/keyman/pull/3817#r517081079_

### PR DESCRIPTION
**Pull request by jahorton**
Fixes #3841.

Also fixes a bug discovered while implementing the changes specified by #3841, which had actually neutered a number of the unit tests.  Fortunately, the only ones that failed when re-enabled... did so because of an error in the test spec, not in the actual engine code.  Whew!

### CODE REVIEW

File: common/core/web/keyboard-processor/tests/cases/engine/context.js [hunk 1]
```diff
@@ -786,73 +892,87 @@ describe('Engine - Context Matching', function() {
 
   describe('handles simple deadkey contexts', function() {
     it('for the most basic cases:  DEADKEY_TEST_1', function() {
-      runEngineRuleSet(DEADKEY_TEST_1, "Deadkeys");
+      runEngineRuleSet([DEADKEY_TEST_1], "Deadkeys");
```
Comment (jahorton) [inline common/core/web/keyboard-processor/tests/cases/engine/context.js:895]: Yeah... this bit right here was kind of important.  The `runEngineRuleSet` method expects an array, but each individual test is merely an object... an object _without length_.

Compare against this part...

https://github.com/keymanapp/keyman/blob/09ad4ce9af5be5695f82d11c02a0b3bf488a959e/common/core/web/keyboard-processor/tests/cases/engine/context.js#L26-L29

As a result, the `for`-loop was instantly shortcut, bypassing the actual test!  No bueno.  At least they'll _actually_ be running now.

File: common/core/web/keyboard-processor/tests/cases/engine/context.js [hunk 2]
```diff
@@ -736,16 +736,122 @@ var DEADKEY_STORE_TEST_4 = {
   }]
 };
 
+var NOTANY_NUL_TEST_1 = {
+  id: 4,
+  // Match condition for rule
+  rule: [{t: 'a', a:['a','b','c'], n: 1},{t:'a',a:['a', 'b', 'c']}],
```
Comment (mcdurdin) [inline common/core/web/keyboard-processor/tests/cases/engine/context.js:742]: Can you add in a comment the equivalent kmn for the rule conditions, e.g. 

```
// store(abc) 'abc'
// notany(abc) any(abc)
```

File: common/core/web/keyboard-processor/tests/cases/engine/context.js [hunk 3]
```diff
@@ -736,16 +736,122 @@ var DEADKEY_STORE_TEST_4 = {
...
+  // Resulting context map
+  contextCache: ['a', 'a'],
+
+  baseSequence: { "output": "aa", "inputs": [
+    {"type":"key","keyCode":65,"states":10752,"modifiers":0,"modifierChanged":false,"isVirtualKey":true}, //a
+    {"type":"key","keyCode":65,"states":10752,"modifiers":0,"modifierChanged":false,"isVirtualKey":true}  //a
+  ]},
+  fullMatchDefs: [{
+    sequence: { "output": "a", "inputs": [
+      {"type":"key","keyCode":50,"states":10752,"modifiers":0,"modifierChanged":false,"isVirtualKey":true}, //a
```
Comment (mcdurdin) [inline common/core/web/keyboard-processor/tests/cases/engine/context.js:785]:
```suggestion
      {"type":"key","keyCode":50,"states":10752,"modifiers":0,"modifierChanged":false,"isVirtualKey":true}, //1
```
??

**Review by mcdurdin — APPROVED** [review]
LGTM. Some minor comments.

One note: it takes a while for me to process the rule conditions and sequences in the tests. I am still not 100% sure I understand why there is a `baseSequence` as well as a `fullMatchDefs`?That reduces my confidence in the correctness of the tests a little. However, I am pretty sure I have read them correctly now and that they are correct.

**Comment by jahorton** [comment]
> LGTM. Some minor comments.
> 
> One note: it takes a while for me to process the rule conditions and sequences in the tests. I am still not 100% sure I understand why there is a `baseSequence` as well as a `fullMatchDefs`?That reduces my confidence in the correctness of the tests a little. However, I am pretty sure I have read them correctly now and that they are correct.

That's... a bit of legacy in the test definition design.

The 'root' of each test definition is an engine rule application that _must_ pass.  Not only is it run during the individual rule's unit test section, but it is _also_ run as one part of many within the test `'properly generates extended context data needed by context-matching checks below'` within the same test-case file.  This (white-box) batch-test ensures that the simulated `baseSequence` generates the expected `contextCache` value when input.

`fullMatchDefs` entries do not run as part of that batch; they don't specify `contextCache` values for comparison.  Those are intended to be variants of the 'base' test that may return either `true` or `false`, checking the same underlying rule behavior with slightly different inputs using a black-box approach.

**Comment by jahorton** [comment]
Thinking about it... perhaps those two different test types could be split off into two different test suites.  It'd almost surely be a clearer... though that might cause a fair bit of test WETness.

File: common/core/web/keyboard-processor/tests/cases/engine/context.js (hunk 2 above)
Comment (jahorton) [reply common/core/web/keyboard-processor/tests/cases/engine/context.js:831]: I've gone ahead and done this for every rule in the suite, even the older ones.

**Comment by mcdurdin** [comment]
> Thinking about it... perhaps those two different test types could be split off into two different test suites. It'd almost surely be clearer... though that might cause a fair bit of test WETness.

I reckon it'd certainly be clearer but it's probably sufficient to just clarify the difference between the two test types somewhere in the source file (whitebox vs blackbox is good info!)

**Pull request merged**